src/metabase/lib/js.cljs

```diff
@@ -418,6 +421,67 @@
   [a-query stage-number column]
   (lib.core/remove-field a-query stage-number column))
 
+(def ^:private visible-columns-defaults
+  {:include-joined?              true
+   :include-expressions?         true
+   :include-implicitly-joinable? true})
+
+
+;; TODO: Added as an expedient to fix metabase/metabase#32373. Due to the interaction with viz-settings, this issue
+;; was difficult to fix entirely within MLv2. Once viz-settings are ported, this function should not be needed, and the
+;; FE logic using it should be ported to MLv2 behind more meaningful names.
+(defn ^:export visible-columns
+  "Return a sequence of column metadatas for columns visible at the given stage of the query. Takes a map of options
+  which defines the set of columns to return. All the options default to *true*, ie. everything is returned by default.
+  ```
+  {
+    include_joined: true,
+    include_expressions true,
+    include_implicitly_joinable: true,
+  }
+  ```"
+  [a-query stage-number ^js js-options]
+  (let [options     (merge visible-columns-defaults
+                           (when-let [opts (some-> js-options js-keys set)]
+                             (cond-> {}
+                               (opts "includeJoined")
+                               (assoc :include-joined? (.-includeJoined js-options))
+                               (opts "includeExpressions")
+                               (assoc :include-expressions? (.-includeExpressions? js-options))
+                               (opts "includeImplicitlyJoinable")
+                               (assoc :include-implicitly-joinable? (.-includeImplicitlyJoinable js-options)))))
+        stage            (lib.util/query-stage a-query stage-number)
+        vis-columns (lib.metadata.calculation/visible-columns a-query stage-number stage options)
+        ret-columns (lib.metadata.calculation/returned-columns a-query stage-number stage)]
+    (to-array (lib.equality/mark-selected-columns vis-columns ret-columns))))
+
+(defn ^:export legacy-field-ref
+  "Given a column metadata from eg. [[fieldable-columns]], return it as a legacy JSON field ref."
+  [column]
+  (-> column
+      lib.core/ref
+      convert/->legacy-MBQL
+      (update 2 update-vals #(if (qualified-keyword? %)
+                               (u/qualified-name %)
+                               %))
+      clj->js))
+
+(defn ^:export find-column-index-from-legacy-ref
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`), search through it for a field matching the
+  given legacy field ref. Returns the index, or -1 if no matching column is found."
```

Comment:
Returning -1 is quite standard (especially in Java and C), but I wonder if returning `nil` would be more comfortable. (The implementation of the function would definitely be a bit more succinct.)

Reply:
It's also conventional in Clojure to return nil rather than empty collections, and to return Clojure collections rather than JS arrays, and we do both those things in this file. This is the JS shim file, and I argue for following JS conventions in here.

Reply:
If JS people expect -1 for not found from such a function, then by all means keep it. 👍